builtin/credential/appgroup/user_id.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	// From the selector field supplied as the credential, detect the type of UserID
+	// supplied. UserID will be verified based on the type.
+	selectorType := ""
+	selectorValue := ""
+	switch {
+	case selector == "generic":
+		selectorType = "generic"
+		selectorValue = b.salt.SaltID(userID)
+	case strings.HasPrefix(selector, "app/") || strings.HasPrefix(selector, "group/"):
+		selectorFields := strings.SplitN(selector, "/", 2)
+		if len(selectorFields) != 2 {
+			return nil, fmt.Errorf("invalid selector; selector type and value could not be parsed")
+		}
+		selectorType = strings.TrimSpace(selectorFields[0])
+		selectorValue = strings.TrimSpace(selectorFields[1])
+		if selectorValue == "" {
+			return nil, fmt.Errorf("missing selector value")
+		}
+	default:
+		return nil, fmt.Errorf("unrecognized selector")
+	}
+
+	// Do the selector validation first. If this results in an error, the UserID
+	// entry should not be modified. Return the validation response if the UserID
+	// is found to be valid and if the UserID entry is updated properly.
+	validationResp, err := b.validateSelector(s, selectorType, selectorValue)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if the user ID supplied is valid. If use limit was specified
+	// on the user ID, it will be decremented in this call.
+	valid, err := b.userIDEntryValid(s, selectorType, selectorValue, userID)
+	if err != nil {
+		return nil, err
+	}
+	if !valid {
+		return nil, fmt.Errorf("user ID not found under the %s selector type", selectorType)
+	}
+
+	return validationResp, nil
+}
+
+// Check if there exists an entry in the name of selectorValue for the selectorType supplied.
+// Prepares a response containing the combined set of policies, TTL and Wrapped values that are
+// applicable to the login.
+func (b *backend) validateSelector(s logical.Storage, selectorType, selectorValue string) (*validationResponse, error) {
+	resp := &validationResponse{
+		SelectorType:  selectorType,
+		SelectorValue: selectorValue,
+	}
+	switch selectorType {
+	case selectorTypeApp:
+		app, err := b.appEntry(s, selectorValue)
+		if err != nil {
+			return nil, err
+		}
+		if app == nil {
+			return nil, fmt.Errorf("app referred by the user ID does not exist")
+		}
+		resp.Policies = app.Policies
+		resp.TokenTTL = app.TokenTTL
+		resp.TokenMaxTTL = app.TokenMaxTTL
+		resp.Wrapped = app.Wrapped
+	case selectorTypeGroup:
+		group, err := b.groupEntry(s, selectorValue)
+		if err != nil {
+			return nil, err
+		}
+		if group == nil {
+			return nil, fmt.Errorf("group referred by the user ID does not exist")
+		}
+		groupPolicies, err := b.fetchPolicies(s, group.Apps)
+		if err != nil {
+			return nil, err
+		}
+		// Append the union of all the policies from all the apps on the group
+		resp.Policies = append(resp.Policies, groupPolicies...)
+
+		// Append the additional policies set on the group
+		resp.Policies = append(resp.Policies, group.AdditionalPolicies...)
+
+		resp.TokenTTL = group.TokenTTL
```

Comment:
The case where this and other things (like wrapped) are not set on a group, or are set differently within different component apps, needs to be figured out. It may be that we simply use whatever is set on a group and the defaults otherwise, but there's an argument for the idea that if an app is more restrictive we should restrict within the group, so that the group doesn't get around restrictions set on individual apps.

This can potentially make it a bit harder to reason about, but the flip side is that it guarantees that you can't escape security by simply asking a hapless operator to make a group for you with the values you give her.

Since any given app could be defined multiple times (with different names obviously) then if you actually intend for app owners to have more restrictive credentials than what goes into a group, you can simply have two apps defined and only put one in the group.